### PR TITLE
mds/cdir: fix the type of dentry in the judgment; cleanup undefined functions

### PR DESCRIPTION
**Pull request by renhwztetecs**
1. Perfect is_primary/is_remote/is_null type of dentry in the judgment;
2. first_get() and last_put() are not defined,so cleanup them

Signed-off-by: huanwen ren <ren.huanwen@zte.com.cn>

### CODE REVIEW

File: src/mds/CDir.cc [hunk 1]
```diff
@@ -608,10 +608,8 @@ void CDir::unlink_inode_work( CDentry *dn )
       dn->unlink_remote(dn->get_linkage());
 
     dn->get_linkage()->set_remote(0, 0);
-  } else {
+  } else if (dn->get_linkage()->is_primary()) {
```
Comment (jcsp) [inline src/mds/CDir.cc]: Do we really ever expect a null dentry here?  If not, then there should be an assert(!dn->get_linkage()->is_null()) after this if block.

File: src/mds/CDir.cc [hunk 2]
```diff
@@ -2187,7 +2185,7 @@ void CDir::_encode_dentry(CDentry *dn, bufferlist& bl,
     bl.append('L');         // remote link
     ::encode(ino, bl);
     ::encode(d_type, bl);
-  } else {
+  } else if (dn->linkage.is_primary()) {
```
Comment (jcsp) [inline src/mds/CDir.cc]: Same as above: if we really never expect a null dentry here, add an assertion before or after the if block

**Comment by jcsp** [comment]
Is this based on any bug you found, or just from code inspection?

File: src/mds/CDir.cc (hunk 1 above)
Comment (renhwztetecs) [inline src/mds/CDir.cc]: null dentry Shouldn't be here，so add assert(!dn->get_linkage()->is_null()) in the else branch

File: src/mds/CDir.cc (hunk 2 above)
Comment (renhwztetecs) [inline src/mds/CDir.cc]: Same as above
thank you

**Comment by renhwztetecs** [comment]
just from code inspection

**Comment by jcsp** [comment]
Realised after testing that the "cleanup undefined functions" commit is causing lots of crashes -- they weren't really unused: please remove that commit.

**Comment by renhwztetecs** [comment]
@jcsp 
I change it later

**Comment by renhwztetecs** [comment]
@jcsp 
remove that commit of cleanup undefined functions

**Comment by jcsp** [comment]
http://pulpito.ceph.com/jspray-2016-06-14_13:37:22-fs-wip-jcsp-testing-20160614-distro-basic-mira/

**Pull request merged**